File: euler10.py
# Найдите сумму всех простых чисел меньше двух миллионов.

from datetime import datetime
from functools import reduce

n = 2_000_000


def eratosfen(number):
    """Решето Эратосфена: генерация простых чисел вплоть до заданного числа"""
    sieve = list(range(number + 1))
    sieve[1] = 0
    for i in sieve:
        if i > 1:
            for j in range(2 * i, len(sieve), i):
                sieve[j] = 0
    return sieve

# 1 Вариант
start_time = datetime.now()
print(reduce(lambda x, y: x + y, eratosfen(n)))
print(datetime.now() - start_time)

# 2 Вариант
start_time = datetime.now()
pass
print(datetime.now() - start_time)

File: go/euler10.go
// euler10 находит сумму всех простых чисел меньше двух миллионов.
//
// Решение задачи Эйлера №10.
package main

import (
	"fmt"
	"time"
)

func eratosfen(number int) []bool {
	sieve := make([]bool, number+1)
	for i := range sieve {
		sieve[i] = true
	}
	sieve[0] = false
	sieve[1] = false
	for i := 2; i*i <= number; i++ {
		if sieve[i] {
			for j := i * i; j <= number; j += i {
				sieve[j] = false
			}
		}
	}
	return sieve
}

func euler10() {
	n := 2000000

	startTime := time.Now()
	sieve := eratosfen(n)
	sum := 0
	for i := 2; i <= n; i++ {
		if sieve[i] {
			sum += i
		}
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	sieve = eratosfen(n)
	sum = 0
	for i := 2; i <= n; i++ {
		if sieve[i] {
			sum += i
		}
	}
	fmt.Println(sum)
	fmt.Println(time.Since(startTime))
}

func main() {
	euler10()
}
